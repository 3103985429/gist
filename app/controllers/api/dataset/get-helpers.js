const winston = require('winston');
const _ = require('lodash');
const path = require('path');

const datasource_description = require('../../../models/descriptions');
const nested = require('./nested');
const indexHelpers = require('./index-helpers');
const uploadHelpers = require('./upload-helpers');

const userSrcPath = path.join(__dirname, '../../../../user/');
const { PublicError } = require('../../../libs/system/errors');
const { restoreColorMapping } = require('../../client/colorPalette');

/**
 * Returns dataset by query.
 *
 * @param {Object} query
 */
const getDatasetByQuery = (query) => new Promise((resolve, reject) => {
    // a preview is a description with only a few fields that are necessary for generating viz preview
    // master_id is a preview description property that points to the full, original description
    if (query.master_id) { //getting the preview copy
        datasource_description.findOne(query, (err, dataset) => {
            if (err) {
                winston.error(`getAllDatasetsWithQuery query: ${query} err: ${err}`);
                return reject({ statusCode: 500, message: err });
            }
            resolve({ statusCode: 200, data: { datasets: dataset } });
        });
    } else {
        if (_.isEmpty(query)) {
            winston.error(`getAllDatasetsWithQuery empty query: ${query}`);
            return reject({ statusCode: 500, message: 'query not found' });
        }
        datasource_description.find({
            $and: [
                { $or: [{ replaced: false }, { replaced: { $exists: false } }] },
                { master_id: { $exists: false }, schema_id: { $exists: false } },
                query,
            ],
        }, {
            _id: 1,
            uid: 1,
            title: 1,
            importRevision: 1,
            sample: 1,
            replaced: 1,
            firstImport: 1,
            replacement: 1,
            child_replacement: 1,
            author: 1,
            createdAt: 1,
            fe_views: 1,
            fe_filters: 1,
        })
            .populate({ path: 'author', select: { firstName: 1, lastName: 1, _id: 1 } })
            .exec((err, datasets) => {
                if (err) {
                    winston.error(`getAllDatasetsWithQuery query: ${query} err: ${err}`);
                    return reject({ statusCode: 500, message: err });
                }
                if (_.isEmpty(datasets)) {
                    return reject({ statusCode: 404, message: 'not found' });
                }
                return resolve({ statusCode: 200, data: { datasets: datasets } });
            });
    }
});
module.exports.getDatasetByQuery = getDatasetByQuery;

/**
 * Find dataset by id ({args.datasetId}) and populate it in {args.description}.
 *
 * @param {object} args
 * @param {function} next
 */
const findDataset = (args, next = _.identity) => new Promise((resolve, reject) => {
    args.populate = ['author', 'schema_id', 'schema_id._team', '_team'];

    datasource_description.findByIdAndPopulateFields(args, (err, description) => {
        if (err) {
            reject(err);
            return next(err);
        }

        args.replacement = description.replacement && description.apiEndPoint;

        if (description.schema_id) {
            description = datasource_description.consolidateDescriptions(description);
        }
        args.description = description;
        resolve(args);
        next(null, args);
    });
});
module.exports.findDataset = findDataset;

/**
 * Reads datasource and gets the list of columns.
 *
 * @param {function} readDatasourceFunction the function to process datasource
 * @param {boolean} replacement
 * @param {DatasourceDescription_scheme} description
 * @param {fs.Stream} stream
 */
const readDatasourceAndSetColumns = (readDatasourceFunction, replacement, description, stream) => new Promise((resolve, reject) => {
    readDatasourceFunction(replacement, description, stream, (err, columns) => {
        if (err) {
            winston.error(err);
            return reject({ statusCode: 500, data: { errorWithData: true, error: err, description: description } });
        }

        // replace the source name and update description columns if it's a replacement
        try {
            if (columns.length === 0) {
                for (let i = 0; i < description.columns.length; i++) {
                    description.columns[i].sourceName = description.apiEndPoint;
                }
            } else {
                const saveColumns = uploadHelpers.setSaveColumnsFunction(description);
                description.raw_rowObjects_coercionScheme = saveColumns(columns);
                description.columns = columns;
            }
            resolve({ statusCode: 200, data: { dataset: description } });
        } catch (e) {
            winston.error(`Error reading datasource columns and sample err: ${e}`);
            description.fileName = undefined;

            reject({
                statusCode: 500,
                data: {
                    errorWithData: true,
                    description: description,
                    error: new PublicError('Error reading datasource columns and sample'),
                },
            });
        }
    });
});
module.exports.readDatasourceAndSetColumns = readDatasourceAndSetColumns;

/**
 * Gets the datasource function.
 *
 * @param {DatasourceDescription_scheme} description
 */
module.exports.getReadDatasourceFunction = (description) => {
    let importCtrl;
    const team = description._team;
    if (!description.permissions || !description.permissions.includes('useStandardImport')) {
        try {
            importCtrl = require(path.join(userSrcPath, team.subdomain, '/src/import'));
            winston.debug(`reading file with ${team.subdomain} import function`);
            if (typeof importCtrl.readDatasourceColumnsAndSampleRecords !== 'undefined') {
                return importCtrl.readDatasourceColumnsAndSampleRecords;
            }
        } catch (err) {
            winston.error(`reading file with index helpers: ${err.message}`);
            return indexHelpers.readDatasourceColumnsAndSampleRecords;
        }
    }
    winston.debug('reading file with index helpers');
    return indexHelpers.readDatasourceColumnsAndSampleRecords;
};

module.exports.updateColorMapping = (args, next) => {
    if (args.description.colorMapping) {
        args.description.colorMapping = restoreColorMapping(args.description.colorMapping);
    }

    next(null, args);
};
