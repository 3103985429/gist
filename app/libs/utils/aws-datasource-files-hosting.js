const winston = require('winston');
const aws = require('aws-sdk');
const fs = require('fs');
const _ = require('lodash');

const bucket = process.env.DO_S3_BUCKET;
const endpoint = new aws.Endpoint(process.env.DO_S3_ENDPOINT);
const s3 = new aws.S3({
    endpoint,
    accessKeyId: process.env.DO_ACCESS_KEY_ID,
    secretAccessKey: process.env.DO_SECRET_ACCESS_KEY,
});
const debugS3 = false;

function _uploadAncillaryFile(filePath, fileName, contentType, teamSubdomin, datasetId, callback) {
    const file = fs.createReadStream(filePath);
    const key = getAncillaryAssetKey(teamSubdomin, datasetId, fileName);
    const params = {
        Bucket: bucket,
        Key: key,
        ContentType: contentType,
        Body: file,
        ACL: 'private',
    };

    debug(`Uploading the ancillary file to S3: ${key}`);

    s3.upload(params, function (err) {
        if (err) {
            return callback(err);
        }

        callback();
    });
}

module.exports.uploadAncillaryFile = _uploadAncillaryFile;

function _copyAncillaryFile(team_subdomain, originId, destinationId, fileName, callback) {
    const key = getAncillaryAssetKey(team_subdomain, destinationId, fileName);
    const params = {
        Bucket: bucket,
        CopySource: getAncillaryAssetKey(team_subdomain, originId, fileName, bucket),
        Key: key,
    };

    debug(`Copying the ancillary file from S3: ${key}`);

    s3.copyObject(params, function (err) {
        if (err) {
            winston.error(`S3: error copying datasource: ${err.message}`);
            return callback(err);
        }

        callback();
    });
}

module.exports.copyAncillaryFile = _copyAncillaryFile;

function _uploadDataSource(filePath, newFilename, contentType, teamSubdomin, datasetId, callback) {
    const file = fs.createReadStream(filePath);
    const key = getDatasourceKey(teamSubdomin, datasetId, newFilename);

    const params = {
        Bucket: bucket,
        Key: key,
        ContentType: contentType,
        Body: file,
        ACL: 'private',
    };

    debug(`Uploading the datasource to S3: ${key}`);

    s3.upload(params, function (err) {
        if (err) {
            return callback(err);
        }

        callback();
    });
}

module.exports.uploadDataSource = _uploadDataSource;

function _copySampleDatasource(datasetToDuplicateId, fileName, datasetId, teamSubdomain, callback) {
    const key = getDatasourceKey(teamSubdomain, datasetId, fileName);
    const params = {
        Bucket: bucket,
        CopySource: getDatasourceKey('sampleteam', datasetToDuplicateId, fileName, bucket),
        Key: key,
    };

    debug(`Copying the sample datasource from S3: ${key}`);

    s3.copyObject(params, function (err) {
        if (err) {
            winston.error(`S3: error copying sample datasource: ${err.message}`);
            return callback(err);
        }

        callback();
    });
}

module.exports.copySampleDatasource = _copySampleDatasource;

function _copyDatasource(team_subdomain, fileName, copyFromId, copyToId, callback) {
    const keyFrom = getDatasourceKey(team_subdomain, copyFromId, fileName, bucket);
    const keyTo = getDatasourceKey(team_subdomain, copyToId, fileName);
    const params = {
        Bucket: bucket,
        CopySource: keyFrom,
        Key: keyTo,
    };

    debug(`Copying the datasource from S3: from ${keyFrom} to ${keyTo}`);

    s3.copyObject(params, function (err, data) {
        if (err) {
            winston.error(`S3: error copying datasource: ${err.message}`);
            return callback(err);
        }

        callback();
    });
}

module.exports.copyDatasource = _copyDatasource;

function _getDatasource(description) {
    const key = getDatasourceKey(description._team.subdomain, description.schemaId || description._id, description.fileName);

    const param = {
        Bucket: bucket,
        Key: key,
    };
    debug(`Reading the datasource from S3: ${key}`);

    return s3.getObject(param);
}

module.exports.getDatasource = _getDatasource;

function _getAncillaryFile(description) {
    const key = getAncillaryAssetKey(description._team.subdomain, description._id, description.ancillaryFile);

    const param = {
        Bucket: bucket,
        Key: key,
    };
    debug(`Reading ancillary file from S3: ${key}`);

    return s3.getObject(param);
}

module.exports.getAncillaryFile = _getAncillaryFile;

function keyExists(description, callback = _.noop) {
    return new Promise((resolve, reject) => {
        const key = getDatasourceKey(description._team.subdomain, description.schema_id || description._id, description.fileName);

        const params = {
            Bucket: bucket,
            Key: key,
        };
        debug(`Verifying the datasource from S3: ${key}`);

        s3.headObject(params, (err) => {
            if (err) {
                reject(err);
                return callback(err, false);
            }

            resolve();
            callback(null, true);
        });
    });
}

module.exports.keyExists = keyExists;

function _deleteDataset(description, callback) {
    const team_subdomain = description._team.subdomain;
    const key = getDatasetKey(team_subdomain, description.id);
    const param = {
        Bucket: bucket,
        Prefix: key,
    };

    debug(`Removing the datasource from S3: ${key}`);

    s3.listObjects(param, function (err, data) {
        if (err) {
            return callback(err);
        }

        if (!data.Contents) {
            return callback();
        }

        if (data.Contents.length === 0) {
            return callback();
        }

        const param = { Bucket: bucket };
        param.Delete = { Objects: [] };

        data.Contents.forEach(function (content) {
            param.Delete.Objects.push({ Key: content.Key });
        });

        s3.deleteObjects(param, function (err, data) {
            if (err) {
                return callback(err);
            }

            if (!data.Deleted) {
                return callback();
            }

            if (data.Deleted.length === 1000) {
                return _deleteDataset(description, callback);
            }

            callback();
        });
    });
}

module.exports.deleteDataset = _deleteDataset;

function _deleteTeam(subdomain, callback) {
    const param = {
        Bucket: bucket,
        Prefix: getTeamKey(subdomain),
    };

    s3.listObjects(param, function (err, data) {
        if (err) {
            return callback(err);
        }

        if (!data.Contents) {
            return callback();
        }

        if (data.Contents.length === 0) {
            return callback();
        }

        const param = { Bucket: bucket };
        param.Delete = { Objects: [] };

        data.Contents.forEach(function (content) {
            param.Delete.Objects.push({ Key: content.Key });
        });

        s3.deleteObjects(param, function (err, data) {
            if (err) {
                return callback(err);
            }

            if (!data.Contents) {
                return callback();
            }

            if (data.Contents.length === 1000) {
                return _deleteTeam(subdomain, callback);
            }

            callback();
        });
    });
}

module.exports.deleteTeam = _deleteTeam;

function _deleteObject(key, callback) {
    const params = {
        Bucket: bucket,
        Key: key,
    };

    s3.deleteObject(params, function (err, data) {
        if (err) {
            return callback(err);
        }

        callback(null, data);
    });
}

module.exports.deleteObject = _deleteObject;

function getDatasourceKey(subdomain, datasetId, fileName, bucket = null) {
    return getDatasetKey(subdomain, datasetId, `datasources/${fileName}`, bucket);
}

function getAncillaryAssetKey(subdomain, datasetId, fileName, bucket = null) {
    return getDatasetKey(subdomain, datasetId, `assets/ancillary/${fileName}`, bucket);
}

function getDatasetKey(subdomain, datasetId, resourceName = '', bucket = null) {
    return getTeamKey(subdomain, `datasets/${datasetId}/${resourceName}`, bucket);
}

function getTeamKey(subdomain, resourceName = '', bucket = null) {
    const prefix = bucket ? `${bucket}/` : '';

    return `${prefix}${subdomain}/${resourceName}`;
}

function debug(msg) {
    if (debugS3) {
        winston.debug(msg);
    }
}
