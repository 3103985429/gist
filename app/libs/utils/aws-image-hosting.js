const request = require('request');
var winston = require('winston');
var aws = require('aws-sdk');
var sharp = require('sharp');
var async = require('async');
const _ = require('lodash');

const { VIEW_TYPE_OBJECT_DETAILS, VIEW_TYPE_GALLERY, VIEW_TYPE_TIMELINE } = require('../../config/views.config');

const BUCKET = process.env.DO_S3_BUCKET;
const endpoint = new aws.Endpoint(process.env.DO_S3_ENDPOINT);
const S3 = new aws.S3({
    endpoint,
    accessKeyId: process.env.DO_ACCESS_KEY_ID,
    secretAccessKey: process.env.DO_SECRET_ACCESS_KEY,
    signatureVersion: 'v4',
    region: 'nyc3',
});

module.exports = {
    uploadToS3,
    signedUrlForPutObject,
    getAllIconsForTeam,
    hostImageLocatedAtRemoteURL,
    copyPageAssets,
    copyDatasetAsset,
    getAssetByKey,
};

function uploadToS3(key, response, callback) {

    var payload = {
        Bucket: BUCKET,
        Key: key,
        Body: response,
        ACL: 'public-read',
    };

    S3.upload(payload, function (err, data) {
        if (err) {
            winston.error('AWS S3 write stream error');
            return callback(err);
        }
        return callback(null, data);
    });
}

function _appendKeyToBucket(key) {
    return 'https://' + BUCKET + '.' + process.env.DO_S3_ENDPOINT + '/' + key;
}

function getImageFormat(headers) {
    const imageFormat = _.get(headers, 'content-type', '').split(';')[0].split('/')[1];

    return imageFormat === 'jpg' ? 'jpeg' : imageFormat;
}

function getImageAtUrl(remoteImageSourceURL, callback) {
    var options = {
        url: remoteImageSourceURL,
        encoding: null,
        timeout: 10000,
    };

    request.get(options, function (err, response, data) {
        if ((err && ['ENOTFOUND', 'ETIMEDOUT'].includes(err.code)) || !response) {
            winston.info(`returning url as null, since Could not read the remote image ${remoteImageSourceURL}: ${err}`);
            return callback(null, null);
        }

        if (err) {
            return callback(err, null);
        }

        const imageFormat = getImageFormat(response.headers);

        if (_.get(sharp, `format.${imageFormat}`)) {
            return callback(null, data);
        }

        return callback(null, null);
    });
}

function signedUrlForPutObject(key, fileType, callback) {
    if (fileType.indexOf('svg') >= 0) { //fix space when uploading svg error
        fileType = 'image/svg+xml';
    }

    var params = {
        Bucket: BUCKET,
        Key: key,
        ACL: 'public-read',
        ContentType: fileType,
    };

    S3.getSignedUrl('putObject', params, function (err, signedUrl) {
        callback(err, { putSignedUrl: signedUrl, publicUrl: _appendKeyToBucket(key) });
    });
}

function getAllIconsForTeam(teamSubdomain, callback) {
    S3.listObjects({
        Bucket: BUCKET,
        Prefix: teamSubdomain + '/assets/icon/',
        Marker: teamSubdomain + '/assets/icon/',
    }, function (err, data) {
        if (err) {
            callback(err);
        } else {
            var listOfUrls = [];
            var objects = data.Contents;

            for (var i = 0; i < objects.length; i++) {
                var objKey = objects[i].Key;

                var url = _appendKeyToBucket(objKey);

                listOfUrls.push(url);
            }
            callback(null, listOfUrls);
        }
    });

}

async function _proceedToStreamToHost(imageUrl, resize, folder, docPKey, callback) {
    try {
        const image = sharp(imageUrl);
        const metadata = await image.metadata();

        if (!metadata) {
            winston.info('returning url as null, since Could not read the remote image metadata');
            return callback();
        }

        const { width, hasAlpha } = metadata;
        const size = width < resize.size ? width : resize.size;
        const whiteColorRGB = { r: 255, g: 255, b: 255 };
        const background = hasAlpha ? { ...whiteColorRGB, alpha: 1 } : whiteColorRGB;
        const resizeConfig = { fit: 'contain', background };
        const resizedImage = resize.cropped ? image.resize(size, size, resizeConfig) : image.resize(size);
        const bufferedImage = await resizedImage.toBuffer();
        const keyName = `${folder}${resize.view}/${docPKey}`;

        uploadToS3(keyName, bufferedImage, callback);
    } catch (error) {
        winston.error('has err: ' + error);
        callback(error);
    }
}

function hostImageLocatedAtRemoteURL(folder, remoteImageSourceURL, overwrite, destinationFilenameSansExt, callback) {
    const recommendedSizes = [
        { size: 763, cropped: false, view: VIEW_TYPE_OBJECT_DETAILS },
        { size: 280, cropped: false, view: VIEW_TYPE_GALLERY },
        { size: 155, cropped: true, view: VIEW_TYPE_TIMELINE },
    ];

    getImageAtUrl(remoteImageSourceURL, function (err, imageBuffer) {
        if (err || !imageBuffer) {
            return callback(err);
        } else {
            async.each(recommendedSizes, function (size, cb) {
                if (overwrite) {
                    _proceedToStreamToHost(imageBuffer, size, folder, destinationFilenameSansExt, cb);
                } else {
                    const keyName = `${folder}${size.view}/${destinationFilenameSansExt}`;
                    const params = {
                        Bucket: BUCKET,
                        Key: keyName,
                    };
                    try {
                        S3.headObject(params, function (err, data) {
                            if (err || !data) {
                                winston.error('Head object err: ' + err);
                            } else {
                                winston.info(' File already uploaded and overwrite is false for object with pKey ' + destinationFilenameSansExt);
                                cb();
                            }
                        });
                    } catch (err) {

                        if (err.code === 'ENOTFOUND' || err.code === 'ETIMEDOUT') {
                            _proceedToStreamToHost(imageBuffer, size, folder, destinationFilenameSansExt, cb);
                        } else {
                            winston.error(err);
                            cb(err);
                        }
                    }
                }
            }, callback);
        }
    });

}

function copyPageAssets(args, callback) {
    var params = {
        Bucket: BUCKET,
        CopySource: BUCKET + '/' + args.subdomain + '/websites/' + args.originalWebsiteId + '/pages/' + args.originalPageId + '/' + args.imageType + '/' + args.fileName,
        Key: args.subdomain + '/websites/' + args.websiteId + '/pages/' + args.pageId + '/' + args.imageType + '/' + args.fileName,
        ACL: 'public-read',
    };

    S3.copyObject(params, function (err) {
        if (err) {
            winston.error('\' error copying page asset with params: \'', params);
            return callback(err);
        } else {
            callback();
        }
    });
}

async function copyDatasetAsset({ fromSubdomain, toSubdomain, fromId, toId, source, fileName }, callback) {
    const getUrl = ({ id, BUCKET, subdomain = fromSubdomain }) =>
        `${BUCKET ? `${BUCKET}/` : ''}${subdomain}/datasets/${id}/${source}/${fileName}`;
    const copySource = getUrl({ id: fromId, BUCKET });
    const key = getUrl({ id: toId, subdomain: toSubdomain });

    const params = {
        Bucket: BUCKET,
        CopySource: copySource,
        Key: key,
        ACL: 'public-read',
    };

    try {
        await S3.copyObject(params).promise();

        callback(null);
    } catch (e) {
        winston.error(`S3: error copying dataset asset: ${e}`);

        callback(e);
    }
}

function getAssetByKey(key, callback) {
    const params = {
        Bucket: BUCKET,
        Key: key,
    };

    S3.getObject(params, function (err, data) {
        if (err) {
            return callback(err);
        } else {
            return callback(null, data);
        }
    });
}
