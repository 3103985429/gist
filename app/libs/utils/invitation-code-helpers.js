const crypto = require('crypto');

/**
 * Returns a sha512 hash of the given string.
 * @param confirmation_code - Code to hash
 * @return
 */
module.exports.createHashWithString = confirmation_code => {
    // Create a new hash, and then update it and return the digest
    const hash = crypto.createHash('sha512');
    hash.update(confirmation_code);
    return hash.digest('hex');
};

/**
 * Generates a random short code.
 * @param codeSize - Size in bytes to generate a code
 * @return
 */
module.exports.generateRandomCode = codeSize => {
    // Get a random buffer
    const randomBuffer = crypto.randomBytes(codeSize);
    // Return the random buffer to string
    return randomBuffer.toString('hex');
};
