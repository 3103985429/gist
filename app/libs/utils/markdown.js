var fs = require('fs');
var marked = require('marked');

module.exports.getMarkdown = function (filePath) {
    var rawMarkdown = fs.readFileSync(filePath, 'utf-8');
    return rawMarkdown;
};

module.exports.parseMarkdown = function (markdown) {
    var options = {
        renderer: new marked.Renderer(),
        gfm: true,
        tables: true,
        breaks: false,
        pedantic: false,
        sanitize: true,
        smartLists: true,
        smartypants: false
    };
    marked.setOptions(options);

    var lexer = new marked.Lexer(options);
    var parsedBody = marked(markdown);
    return parsedBody;
};

module.exports.replaceCharts = function (parsed) {
    var chartRE = /((?:<p>)?&lt;chart data-source-key=&quot;(?=\S*)([a-zA-Z\d_-]*)&quot; data-chart-type=&quot;(?=\S*)([a-zA-Z\d-]*)&quot; class=&quot;chart&quot; data-revision=&quot;(\d+)&quot; \/&gt;(?:<\/p>)?)/g;
    var parsedReplaced = parsed.replace(chartRE, '<div id="$2" data-source-key="$2" data-chart-type="$3" class="$3 chart" data-revision="$4"></div>');
    return parsedReplaced;
};
