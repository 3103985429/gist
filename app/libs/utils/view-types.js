const CONTENT_ONLY_TYPE = 'contentOnly';

module.exports = {
    CONTENT_ONLY_TYPE,
};
