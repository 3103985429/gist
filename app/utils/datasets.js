const { findDataset } = require('../controllers/api/dataset/get-helpers');
const { some, includes } = require('lodash');
const DATA_SOURCE_TYPES = require('../../config/data-source-types');

module.exports.getDataset = getDataset;
module.exports.hasValidDatasetSource = hasValidDatasetSource;

/**
 * @param {String|ObjectId} id
 * @returns {Promise<Object>}
 */
async function getDataset(id) {
    const { description } = await findDataset({ id });

    return description;
}

/**
 * Determins if the datasourceDescription passed has a valid dataset source
 * determined from the list in config/data-source-types.js
 * @param {Object} datasourceDescription
 * @returns {Boolean}
 */
function hasValidDatasetSource(datasourceDescription) {
    return some(datasourceDescription, (value, key) => {
        return includes(DATA_SOURCE_TYPES, key);
    });
}
