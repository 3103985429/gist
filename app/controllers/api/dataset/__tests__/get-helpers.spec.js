import * as getHelpers from '../get-helpers';
import { hydrateDb } from '../../../../../internals/testing/backend/utils';
import { pokemon1 } from '../../../../../internals/testing/backend/fixtures/datasets';


describe('Get helpers', () => {
    beforeEach(async () => {
        await hydrateDb();
    });

    describe('getDatasetWithQuery', () => {
        it('should return dataset with title', async () => {
            const dataset = await getHelpers.getDatasetByQuery({ title: pokemon1.title });
            expect(dataset.statusCode).toBe(200);
            expect(dataset).toMatchSnapshot();
        });
    });
});
