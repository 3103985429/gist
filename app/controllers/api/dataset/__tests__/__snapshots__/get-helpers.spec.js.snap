// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get helpers getDatasetWithQuery should return dataset with title 1`] = `
Object {
  "data": Object {
    "datasets": Array [
      Object {
        "_id": "5a42c26e303770207ce8f83b",
        "author": Object {
          "_id": "5a42bf07d3bc0b20648290e0",
          "firstName": "Maitland",
          "lastName": "McConnell",
        },
        "createdAt": 2017-12-26T21:43:10.707Z,
        "fe_filters": Object {
          "default": Object {},
          "excludeFields": Array [],
          "fabricated": Array [],
          "fieldsAsRangeSlider": Array [],
          "fieldsCommaSeparatedAsIndividual": Array [],
          "fieldsMultiSelectable": Array [],
          "fieldsNotAvailable": Array [],
          "fieldsSortableByInteger": Array [],
          "fieldsSortableInReverseOrder": Array [],
          "keywords": Array [],
          "valuesToExcludeByOriginalKey": Object {
            "_all": Array [],
          },
        },
        "fe_views": Object {
          "default_view": "gallery",
          "views": Object {
            "gallery": Object {
              "defaultSortByColumnName": "Image URL",
              "visible": true,
            },
            "pieChart": Object {
              "defaultGroupByColumnName": "Name",
              "visible": true,
            },
            "pieSet": Object {
              "defaultAggregateByColumnName": "Max HP",
              "defaultChartByColumnName": "Type 1",
              "defaultGroupByColumnName": "Type 2",
              "fieldsNotAvailableAsAggregateByColumns": Array [],
              "fieldsNotAvailableAsChartByColumns": Array [],
              "fieldsNotAvailableAsGroupByColumns": Array [],
              "valuesToExcludeByOriginalKey": Array [],
              "visible": true,
            },
            "timeline": Object {
              "visible": true,
            },
            "treemap": Object {
              "defaultAggregateByColumnName": "Max HP",
              "defaultChartByColumnName": "Name",
              "defaultGroupByColumnName": "Type 1",
              "fieldsNotAvailableAsAggregateByColumns": Array [],
              "fieldsNotAvailableAsChartByColumns": Array [],
              "fieldsNotAvailableAsGroupByColumns": Array [],
              "visible": true,
            },
          },
        },
        "firstImport": 0,
        "importRevision": 1,
        "replaced": false,
        "replacement": false,
        "sample": false,
        "title": "Pokemon",
        "uid": "pokemon",
      },
      Object {
        "_id": "5b033d8d2a54a24d161a9c9b",
        "createdAt": 2017-12-26T21:43:10.707Z,
        "fe_filters": Object {
          "default": Object {},
          "excludeFields": Array [],
          "fabricated": Array [],
          "fieldsAsRangeSlider": Array [],
          "fieldsCommaSeparatedAsIndividual": Array [],
          "fieldsMultiSelectable": Array [],
          "fieldsNotAvailable": Array [],
          "fieldsSortableByInteger": Array [],
          "fieldsSortableInReverseOrder": Array [],
          "keywords": Array [],
          "valuesToExcludeByOriginalKey": Object {
            "_all": Array [],
          },
        },
        "fe_views": Object {
          "default_view": "gallery",
          "views": Object {
            "gallery": Object {
              "defaultSortByColumnName": "Image URL",
              "visible": true,
            },
          },
        },
        "firstImport": 0,
        "importRevision": 1,
        "replaced": false,
        "replacement": false,
        "sample": false,
        "title": "Pokemon",
        "uid": "pokemon",
      },
    ],
  },
  "statusCode": 200,
}
`;
